File: webapp/functions/_lib/rewrite.js
const ASSET_ATTRS = {
  link: ["href"],
  script: ["src"],
  img: ["src"],
  source: ["src", "srcset"],
  video: ["src", "poster"],
  audio: ["src"],
  embed: ["src"],
  object: ["data"],
  input: ["src"],
};

const SKIP_PREFIXES = ["#", "mailto:", "tel:", "javascript:", "data:"];

function isExternal(path) {
  return /^[a-zA-Z][a-zA-Z0-9+.-]*:/.test(path) || path.startsWith("//");
}

export function classifyHref(href, pageSlugs) {
  href = (href || "").trim();
  if (!href || SKIP_PREFIXES.some((p) => href.startsWith(p))) return { kind: "skip" };
  if (isExternal(href)) return { kind: "external" };

  let url;
  try {
    url = new URL(href, "http://templify.local/");
  } catch {
    return { kind: "skip" };
  }
  const path = url.pathname.replace(/^\/+/, "");
  if (!path) return { kind: "skip" };

  const base = path.split("/").pop();
  const dot = base.lastIndexOf(".");
  const stem = dot >= 0 ? base.slice(0, dot) : base;
  const ext = dot >= 0 ? base.slice(dot).toLowerCase() : "";

  if (ext === ".html" && pageSlugs.has(stem)) {
    return { kind: "page", slug: stem, fragment: url.hash ? url.hash.slice(1) : "" };
  }
  return { kind: "static", relpath: path };
}

export function staticTag(appName, relpath) {
  return `{% static '${appName}/${relpath}' %}`;
}

export function urlTag(appName, slug) {
  return `{% url '${appName}:${slug}' %}`;
}

function rewriteAssetAttr(el, attr, appName) {
  const value = el.getAttribute(attr);
  if (!value || isExternal(value) || SKIP_PREFIXES.some((p) => value.startsWith(p))) return;
  const relpath = value.replace(/^[./]+/, "");
  el.setAttribute(attr, staticTag(appName, relpath));
}

function rewritePageLink(el, attr, appName, pageSlugs) {
  const href = el.getAttribute(attr);
  if (href === null) return;
  const c = classifyHref(href, pageSlugs);
  if (c.kind === "page") {
    let newHref = urlTag(appName, c.slug);
    if (c.fragment) newHref += "#" + c.fragment;
    el.setAttribute(attr, newHref);
  } else if (c.kind === "static") {
    el.setAttribute(attr, staticTag(appName, c.relpath));
  }
}

export function rewriteAssets(document, appName, pageSlugs) {
  for (const [tagName, attrs] of Object.entries(ASSET_ATTRS)) {
    for (const el of document.querySelectorAll(tagName)) {
      for (const attr of attrs) {
        if (el.hasAttribute(attr)) rewriteAssetAttr(el, attr, appName);
      }
    }
  }

  for (const el of document.querySelectorAll("a")) {
    if (el.hasAttribute("href")) rewritePageLink(el, "href", appName, pageSlugs);
  }
}

export function applyNavActiveState(body, appName, pageSlugs) {
  for (const nav of body.querySelectorAll("nav")) {
    for (const a of nav.querySelectorAll("a[href]")) {
      const href = a.getAttribute("href");
      let matched = null;
      for (const slug of pageSlugs) {
        if (href === urlTag(appName, slug)) {
          matched = slug;
          break;
        }
      }
      if (!matched) continue;

      const classes = (a.getAttribute("class") || "").split(/\s+/).filter((c) => c && c !== "active");
      classes.push(`{% if request.resolver_match.url_name == '${matched}' %}active{% endif %}`);
      a.setAttribute("class", classes.join(" "));
    }
  }
}

File: templify/rewrite.py
import os
from urllib.parse import urlsplit

ASSET_ATTRS = {
    "link": ["href"],
    "script": ["src"],
    "img": ["src"],
    "source": ["src", "srcset"],
    "video": ["src", "poster"],
    "audio": ["src"],
    "embed": ["src"],
    "object": ["data"],
    "input": ["src"],
}

SKIP_PREFIXES = ("#", "mailto:", "tel:", "javascript:", "data:")


def _is_external(path):
    parts = urlsplit(path)
    return bool(parts.scheme) or path.startswith("//")


def classify_href(href, page_slugs):
    href = (href or "").strip()
    if not href or href.startswith(SKIP_PREFIXES):
        return ("skip", href)
    if _is_external(href):
        return ("external", href)

    parts = urlsplit(href)
    if not parts.path:
        return ("skip", href)

    base = parts.path.rsplit("/", 1)[-1]
    stem, ext = os.path.splitext(base)
    if ext.lower() == ".html" and stem in page_slugs:
        return ("page", stem, parts.fragment)
    return ("static", parts.path.lstrip("./"))


def static_tag(app_name, relpath):
    return "{%% static '%s/%s' %%}" % (app_name, relpath)


def url_tag(app_name, page_slug):
    return "{%% url '%s:%s' %%}" % (app_name, page_slug)


def rewrite_asset_attr(tag, attr, app_name):
    value = tag.get(attr)
    if not value or _is_external(value) or value.startswith(SKIP_PREFIXES):
        return
    relpath = value.lstrip("./")
    tag[attr] = static_tag(app_name, relpath)


def rewrite_page_link(tag, attr, app_name, page_slugs):
    href = tag.get(attr)
    if href is None:
        return
    kind, *rest = classify_href(href, page_slugs)
    if kind == "page":
        slug, fragment = rest
        new_href = url_tag(app_name, slug)
        if fragment:
            new_href += "#" + fragment
        tag[attr] = new_href
    elif kind == "static":
        (relpath,) = rest
        tag[attr] = static_tag(app_name, relpath)


def rewrite_assets(soup, app_name, page_slugs):
    for tag_name, attrs in ASSET_ATTRS.items():
        for tag in soup.find_all(tag_name):
            for attr in attrs:
                if tag.has_attr(attr):
                    rewrite_asset_attr(tag, attr, app_name)

    for tag in soup.find_all("a"):
        if tag.has_attr("href"):
            rewrite_page_link(tag, "href", app_name, page_slugs)


def apply_nav_active_state(container, app_name, page_slugs):
    for nav in container.find_all("nav"):
        for a in nav.find_all("a", href=True):
            href = a["href"]
            matched = None
            for slug in page_slugs:
                if href == url_tag(app_name, slug):
                    matched = slug
                    break
            if matched is None:
                continue
            classes = [c for c in a.get("class", []) if c != "active"]
            classes.append(
                "{%% if request.resolver_match.url_name == '%s' %%}active{%% endif %%}" % matched
            )
            a["class"] = " ".join(classes)
